### PR TITLE
Client Accounts.forgotPassword and .verifyEmail pass error in callback

### PR DESCRIPTION
**Pull request by laosb**
Issue #5664 
Previous PR #5681

This rebased @mishavee 's work. Thank you @mishavee !

### CODE REVIEW

**Comment by laosb** [comment]
Failing this: 
```
C: tinytest - passwords - changing passwords : !!!!!!!!! FAIL !!!!!!!!!!!
TypeError: 'undefined' is not a function (evaluating 'callback(new Meteor.Error(400, "Password may not be empty"))')
    at http://localhost:4096/packages/accounts-password.js?hash=862511339ab24b4e1e7ce2ff8c7c7eb124812a89:190
    at http://localhost:4096/packages/local-test_accounts-password.js?hash=cfdd62289298c8f1242c7288d79099ae3ba32d46:405
    at http://localhost:4096/packages/test-helpers.js?hash=be3b6e8a6205a0152f0b4544ad2896638c07334e:273
    at http://localhost:4096/packages/test-helpers.js?hash=be3b6e8a6205a0152f0b4544ad2896638c07334e:260
    at http://localhost:4096/packages/test-helpers.js?hash=be3b6e8a6205a0152f0b4544ad2896638c07334e:237
    at http://localhost:4096/packages/test-helpers.js?hash=be3b6e8a6205a0152f0b4544ad2896638c07334e:221
    at http://localhost:4096/packages/test-helpers.js?hash=be3b6e8a6205a0152f0b4544ad2896638c07334e:281
    at http://localhost:4096/packages/test-helpers.js?hash=be3b6e8a6205a0152f0b4544ad2896638c07334e:260
    at http://localhost:4096/packages/test-helpers.js?hash=be3b6e8a6205a0152f0b4544ad2896638c07334e:237
    at http://localhost:4096/packages/test-helpers.js?hash=be3b6e8a6205a0152f0b4544ad2896638c07334e:214
    at http://localhost:4096/packages/accounts-password.js?hash=862511339ab24b4e1e7ce2ff8c7c7eb124812a89:215
    at http://localhost:4096/packages/meteor.js?hash=ed600264974381dc7fe10e24064e3d8c8a8eaa14:1105
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:3541
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:3561
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:4677
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:3369
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:2736
    at http://localhost:4096/packages/underscore.js?hash=27b3d669b418de8577518760446467e6ff429b1e:152
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:2737
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:175
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:1160
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:1218
    at http://localhost:4096/packages/ddp-client.js?hash=40a6742de1fccb42fc4e71f0eb5349f2df33c3c8:1365
```

**Comment by laosb** [comment]
My fault. Rebuilding.

**Comment by laosb** [comment]
Added in History.md.

**Comment by laosb** [comment]
Pinging @stubailo who reviewed the old PR.

**Comment by Menda** [comment]
Needs rebasing

**Comment by laosb** [comment]
Yeah, will do this weekend.

**Comment by laosb** [comment]
God slow Internet connection... Will try if I have a good connection.

**Comment by benjamn** [comment]
I would feel much better about these changes if we continued throwing the exception when there is no `callback`, rather than silently ignoring it.

**Comment by tmeasday** [comment]
@laosb to follow up on @benjamn's comment: although this is a well established pattern in this file, it seems really wrong to silently return from these functions when the **OPTIONAL** callback is not provided and the arguments are wrong.

Can we add something to this file like 

```js
reportError(error, callback) {
   if (callback) {
     callback(error);
     return false;
   } else {
     throw error;
   }
}
```

And replace *all* the instances of

```
callback && callback(...);
return;
```

With

```
return reportError(callback, ...)
````

**Comment by laosb** [comment]
Oh yeah of course.

**Comment by tmeasday** [comment]
@laosb were you planning on looking at this?

**Comment by laosb** [comment]
@tmeasday Sorry for the delay! Tests running.

**Comment by tmeasday** [comment]
No problem @laosb !

**Comment by laosb** [comment]
My fault of copy&pasting. But the tests shouldn't hang on running when there is a error?
image

**Comment by laosb** [comment]
@tmeasday Code fixed but seems that tests are not suitable for this update. Can you give me some hints on modifying tests?

File: packages/accounts-password/password_client.js [hunk 1]
```diff
@@ -118,8 +125,7 @@ Accounts.createUser = function (options, callback) {
   if (typeof options.password !== 'string')
     throw new Error("options.password must be a string");
```
Comment (tmeasday) [inline packages/accounts-password/password_client.js]: Should we use `reportError` on this line too?

Comment (tmeasday) [inline packages/accounts-password/password_client.js]: (Maybe not, we just call `check` in other places)

File: packages/accounts-password/password_tests.js [hunk 2]
```diff
@@ -410,6 +410,13 @@ if (Meteor.isClient) (function () {
       "email", [
     createUserStep,
     logoutStep,
+    // Create user error without callback should not throw error
+    function (test, expect) {
```
Comment (tmeasday) [inline packages/accounts-password/password_tests.js]: Why did you add this new step? I think other tests cover this situation, no? Plus it fails because the password is empty.

File: packages/accounts-password/password_tests.js [hunk 3]
```diff
@@ -468,6 +475,14 @@ if (Meteor.isClient) (function () {
         test.equal(Meteor.user().username, self.username);
       }));
     },
+    // change password with bad old password, but no callback so no error
+    function (test, expect) {
+      Accounts.changePassword('wrong', 'doesntmatter');
```
Comment (tmeasday) [inline packages/accounts-password/password_tests.js]: These should now throw an error right?

File: packages/accounts-password/password_tests.js [hunk 4]
```diff
@@ -544,6 +559,74 @@ if (Meteor.isClient) (function () {
...
+    // forgotPassword called on client with blank email
+    function (test, expect) {
+      Accounts.forgotPassword(
+        { email: this.email }, expect(function (error) {
+          test.isTrue(error);
+      }));
+    },
+    // forgotPassword called on client with blank email, no callback so no error
+    function (test, expect) {
+      Accounts.forgotPassword({ email: this.email });
```
Comment (tmeasday) [inline packages/accounts-password/password_tests.js]: Again, this is no longer true.

**Comment by tmeasday** [comment]
@laosb I made some comments. It looks mostly like we had / you added tests that are now failing because the behaviour is stricter than before. We should just fix those tests to expect errors to be thrown (using `test.throws`) and I think we'll be good.

**Comment by laosb** [comment]
Not sure about what's happening on circle ci. @tmeasday

**Comment by abernix** [comment]
@laosb I would probably rebase this against `devel`, to start with; a lot has changed in `devel` since it was rebased last.

**Comment by laosb** [comment]
@abernix Yeah I think we should do that. Is it possible that the CircleCI results are related to outdate?

**Comment by abernix** [comment]
That's just the first step to solving CI bugs – especially when it seems that everything is failing in the way it is, you might have been half-way between commits or something when this was originally branched.

**Comment by tmeasday** [comment]
Looks like there's one failing test left @laosb!

**Comment by laosb** [comment]
@tmeasday Yeah, but I don't know how to fix it. The code does throws an error according to the log, but the test suite complains about no error was thrown.

**Comment by laosb** [comment]
Oops, my fault of using `pull`...

**Comment by tmeasday** [comment]
@laosb I would try rebasing your commits -- something like

```
git fetch
git checkout your-branch
git reset --hard <last-commit-hash-of-yours>
git rebase <commit-where-you-branched-off-devel> --onto origin/devel
```

**Comment by laosb** [comment]
Ok, just rebased!

**Comment by laosb** [comment]
```
Exception in delivering result of invoking 'changePassword': Error: Incorrect password [403]
    at reportError (http://localhost:4096/packages/accounts-password.js?hash=2509f4ba185764fc21922993e746e175ed4a40c3:143)
    at http://localhost:4096/packages/accounts-password.js?hash=2509f4ba185764fc21922993e746e175ed4a40c3:228
    at http://localhost:4096/packages/meteor.js?hash=ed600264974381dc7fe10e24064e3d8c8a8eaa14:1105
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:3557
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:3577
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:4738
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:3385
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:2736
    at http://localhost:4096/packages/underscore.js?hash=27b3d669b418de8577518760446467e6ff429b1e:152
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:2737
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:175
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:1160
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:1218
    at http://localhost:4096/packages/ddp-client.js?hash=2da97e4abeeb8a6d368dcaee62429eca9ff5faca:1365
C: tinytest - passwords - changing passwords : !!!!!!!!! FAIL !!!!!!!!!!!
{"groupPath":["tinytest","passwords"],"test":"changing passwords","events":[{"sequence":110,"type":"fail","details":{"type":"throws","message":"did not throw an error as expected","asyncBlock":4},"cookie":{"name":"passwords - changing passwords","offset":0,"groupPath":["tinytest","passwords"],"shortName":"changing passwords"}}]}
```

According to the log, it did threw an error but the test suites didn't seen it. @tmeasday can you take a look whether I did something wrong?

**Comment by tmeasday** [comment]
@laosb the reasons that test is failing is that the error is thrown in a callback (i.e. after the server returns a result of the method call). There's no way on the client to have this thrown as an error you can `try {} catch`, which is why it ends up getting logged to the console and the test fails. 

I think it's OK in this case to not include a test for that behaviour. I've removed the test and I'm merging the PR!

**Comment by tmeasday** [comment]
Closing in favour of https://github.com/meteor/meteor/pull/7695
